opal/mca/threads/qthreads/threads_qthreads_mutex.h

```diff
@@ -42,46 +43,67 @@
 
 BEGIN_C_DECLS
 
-typedef opal_atomic_lock_t opal_thread_internal_mutex_t;
+typedef qthread_spinlock_t opal_thread_internal_mutex_t;
 
-#define OPAL_THREAD_INTERNAL_MUTEX_INITIALIZER           OPAL_ATOMIC_LOCK_INIT
-#define OPAL_THREAD_INTERNAL_RECURSIVE_MUTEX_INITIALIZER OPAL_ATOMIC_LOCK_INIT
+#define OPAL_THREAD_INTERNAL_MUTEX_INITIALIZER           QTHREAD_MUTEX_INITIALIZER
+#define OPAL_THREAD_INTERNAL_RECURSIVE_MUTEX_INITIALIZER QTHREAD_RECURSIVE_MUTEX_INITIALIZER
 
 static inline int opal_thread_internal_mutex_init(opal_thread_internal_mutex_t *p_mutex,
                                                   bool recursive)
 {
-    opal_atomic_lock_init(p_mutex, 0);
+    opal_threads_ensure_init_qthreads();
+    #if OPAL_ENABLE_DEBUG
+    int ret = qthread_spinlock_init(p_mutex,recursive);
+    if (QTHREAD_SUCCESS != ret) {
+        opal_output(0, "opal_thread_internal_mutex_init()");
+    }
+    #else
+    qthread_spinlock_init(p_mutex,recursive);
+    #endif
     return OPAL_SUCCESS;
 }
 
 static inline void opal_thread_internal_mutex_lock(opal_thread_internal_mutex_t *p_mutex)
 {
     opal_threads_ensure_init_qthreads();
-
-    int ret = opal_atomic_trylock(p_mutex);
-    while (0 != ret) {
-        qthread_yield();
-        ret = opal_atomic_trylock(p_mutex);
+    #if OPAL_ENABLE_DEBUG
+    int ret = qthread_spinlock_lock(p_mutex);
+    if (QTHREAD_SUCCESS != ret) {
+        opal_output(0, "opal_thread_internal_mutex_lock()");
     }
+    #else
+    qthread_spinlock_lock(p_mutex);
+    #endif
 }
 
 static inline int opal_thread_internal_mutex_trylock(opal_thread_internal_mutex_t *p_mutex)
 {
     opal_threads_ensure_init_qthreads();
-
-    int ret = opal_atomic_trylock(p_mutex);
-    if (0 != ret) {
-        /* Yield to avoid a deadlock. */
-        qthread_yield();
+    int ret = qthread_spinlock_trylock(p_mutex);
+    if (QTHREAD_OPFAIL == ret) {
+        return 1;
+    } else if (QTHREAD_SUCCESS != ret) {
+#if OPAL_ENABLE_DEBUG
+        opal_output(0, "opal_thread_internal_mutex_trylock()");
+#endif
+        return 1;
+    } else {
```

Comment:
what is the purpose of this `else`? You can write this code without it to increase the readability. Same comment for most of the functions below.

Reply:
Good point, removing the branch.